src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,400 @@
+            foreach (string maybeMatch in Directory.EnumerateFiles(storePath, cert.Thumbprint + PfxWildcard))
+            {
+                hadCandidates = true;
+
+                try
+                {
+                    using (X509Certificate2 candidate = new X509Certificate2(maybeMatch))
+                    {
+                        if (candidate.Equals(cert))
+                        {
+                            return maybeMatch;
+                        }
+                    }
+                }
+                catch (CryptographicException)
+                {
+                    // Contents weren't interpretable as a certificate, so it's not a match.
+                }
+            }
+
+            return null;
+        }
+
+        private string FindOpenSlot(string thumbprint)
+        {
+            // We already know that {thumbprint}.pfx is taken, so start with {thumbprint}.1.pfx
+
+            // We need space for {thumbprint} (thumbprint.Length)
+            // And ".0.pfx" (6)
+            // If MaxSaveAttempts is big enough to use more than one digit, we need that space, too (MaxSaveAttempts / 10)
+            StringBuilder pathBuilder = new StringBuilder(thumbprint.Length + PfxOrdinalWildcard.Length + (MaxSaveAttempts / 10));
+            HashSet<string> existingFiles = new HashSet<string>(Directory.EnumerateFiles(_storePath, thumbprint + PfxOrdinalWildcard));
+
+            for (int i = 1; i <= MaxSaveAttempts; i++)
+            {
+                pathBuilder.Clear();
+
+                pathBuilder.Append(thumbprint);
+                pathBuilder.Append('.');
+                pathBuilder.Append(i);
+                pathBuilder.Append(PfxExtension);
+
+                string builtPath = pathBuilder.ToString();
+
+                if (!existingFiles.Contains(builtPath))
+                {
+                    return Path.Combine(_storePath, builtPath);
+                }
+            }
+
+            throw new CryptographicException(SR.Cryptography_X509_StoreNoFileAvailable);
+        }
+
+        private void DirectoryChanged(object sender, FileSystemEventArgs e)
+        {
+            
+            // Events might end up not firing until after the object was disposed, which could cause
+            // problems consistently reading _watcher; so save it to a local.
+            FileSystemWatcher watcher = _watcher;
+
+            if (watcher != null)
+            {
+                // Stop processing events until we read again, particularly because
+                // there's nothing else we'll do until then.
+                watcher.EnableRaisingEvents = false;
+            }
+
+            _certificates = null;
+        }
+
+        private static string GetDirectoryName(string storeName)
+        {
+            Debug.Assert(storeName != null);
+
+            try
+            {
+                string fileName = Path.GetFileName(storeName);
+
+                if (!StringComparer.Ordinal.Equals(storeName, fileName))
```

Comment:
I find this way easier to read, and it has the benefit of dealing with nulls (which doesn't apply here... but it drives my motor memory)